File: C++/Binary_Tree_00.cpp
// Build a Binary Tree with preorder sequence

// Preorder Sequence : {1 ,2 ,-1 , -1, 3 ,4 ,-1 , -1, 5 , -1, -1}

// In Preorder Traversal : Root , Left , Right
// In Inorder Traversal : Left , Root , Right
// In Postorder Traversal : Left , Right , Root 

#include<iostream>

using namespace std;

class Node{
    public:
    int val;
    Node* left;
    Node* right;

    Node(int data){
        val=data;
        left=NULL;
        right=NULL;
    }
};

static int idx=-1;

Node* Build_Tree(int preorder[]){

    idx++;

    if(preorder[idx]==-1){
        return NULL;
    }

    Node* root=new Node(preorder[idx]);
    root->left=Build_Tree(preorder);
    root->right=Build_Tree(preorder);

    return root;
}

void Preorder_Traversal(Node* Root){
    if(Root==NULL){
        return;
    }

    cout << Root->val << " ";
    Preorder_Traversal(Root->left);
    Preorder_Traversal(Root->right);
}


void Inorder_Traversal(Node* Root){
    if(Root==NULL){
        return;
    }

    Inorder_Traversal(Root->left);
    cout << Root->val << " ";
    Inorder_Traversal(Root->right);
}


void Postorder_Traversal(Node* Root){
    if(Root==NULL){
        return;
    }

    Postorder_Traversal(Root->left);
    Postorder_Traversal(Root->right);
    cout << Root->val << " ";
}


int main(){
    int Preorder[]={1 ,2 ,-1 , -1, 3 ,4 ,-1 , -1, 5 , -1, -1};

    Node* Root=Build_Tree(Preorder);

    cout << Root->val << endl;
    cout << Root->left->val << endl;
    cout << Root->right->val << endl;

    Preorder_Traversal(Root);
    cout << endl;
    Inorder_Traversal(Root);
    cout << endl;
    Postorder_Traversal(Root);
    cout << endl;

    return 0;
}

File: Java/Binary_Tree_00.java
public class Binary_Tree_00 {

    static class Node {
        int data;
        Node left;
        Node right;

        Node(int data) {
            this.data = data;
            left = null;
            right = null;
        }
    }

    static class Binary_Tree {
        int idx = -1;

        // Build Tree from given Preorder Sequence
        public Node Build_Tree(int[] Preorder) {
            idx++;

            if (Preorder[idx] == -1) {
                return null;
            }

            Node new_Node = new Node(Preorder[idx]);
            new_Node.left = Build_Tree(Preorder);
            new_Node.right = Build_Tree(Preorder);

            return new_Node;
        }

        // Preorder_Traversal
        public void Preorder_Traversal(Node Root) {
            if (Root == null) {
                return;
            }

            System.out.print(Root.data + " ");
            Preorder_Traversal(Root.left);
            Preorder_Traversal(Root.right);
        }

        // Inorder_Traversal
        public void Inorder_Traversal(Node Root) {
            if (Root == null) {
                return;
            }

            Preorder_Traversal(Root.left);
            System.out.print(Root.data + " ");
            Preorder_Traversal(Root.right);
        }

        // Postorder_Traversal
        public void Postorder_Traversal(Node Root) {
            if (Root == null) {
                return;
            }

            Preorder_Traversal(Root.left);
            Preorder_Traversal(Root.right);
            System.out.print(Root.data + " ");
        }
    }

    public static void main(String[] args) {
        int[] Preorder = { 1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1 };

        Binary_Tree tree = new Binary_Tree();

        Node Root = tree.Build_Tree(Preorder);

        System.out.println(Root.data);

        System.out.println("Preorder_Traversal: ");
        tree.Preorder_Traversal(Root);
        System.out.println();

        System.out.println("Inorder_Traversal: ");
        tree.Inorder_Traversal(Root);
        System.out.println();

        System.out.println("Postorder_Traversal: ");
        tree.Postorder_Traversal(Root);
        System.out.println();
    }
}
